File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> map = new HashMap<>();
        for(int num:nums) map.put(num, map.getOrDefault(num,0)+1);
        int n = nums.length;
        List<Integer> bucket[] = new List[n+1];
        for(int key : map.keySet())
        {
            int freq = map.get(key);
            if(bucket[freq]==null) bucket[freq] = new ArrayList<>();
            bucket[freq].add(key);
        }
        int result[] = new int[k];
        int idx = 0;
        for(int i = bucket.length-1;i>=0 && idx<k;i--)
        {
            if(bucket[i]!=null)
            {
                for(int num:bucket[i])
                {
                    result[idx++] = num;
                    if(idx == k) break;
                }
            }
        }
        return result;
    }
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = {}
        for num in nums:
            freq[num] = freq.get(num,0)+1
        heap = []
        for num in freq:
            heapq.heappush(heap,(freq[num],num))
            if len(heap)>k:
                heapq.heappop(heap)
        res = []
        for f,num in heap:
            res.append(num)
        return res